libraries/CurieBLE/src/internal/BLEDeviceManager.cpp

```diff
@@ -336,109 +336,131 @@ void BLEDeviceManager::setAppearance(unsigned short appearance)
 }
 
 BLE_STATUS_T
+BLEDeviceManager::setAdvertiseData(uint8_t type, const uint8_t* data, uint8_t length)
+{
+    uint8_t lengthOfAdv = 0; // Flags data length
+    uint8_t lengthOfScanRsp = 0; // Flags data length
+    bt_data_t *fill_area = NULL;
+    
+    // Get the length of the Advertisement
+    for (uint8_t i = 0; i < _adv_data_idx; i++)
+    {
+        lengthOfAdv += _adv_data[i].data_len + 2;
+    }
+    
+    for (uint8_t i = 0; i < _scan_rsp_data_idx; i++)
+    {
+        lengthOfAdv += _scan_rsp_data[i].data_len + 2;
+    }
+        
+        
+    if (((length + lengthOfAdv) < BLE_MAX_ADV_SIZE) && 
+        (_adv_data_idx < ARRAY_SIZE(_adv_data)))
+    {
+        fill_area = &_adv_data[_adv_data_idx];
+        _adv_data_idx++;
+    }
+    else if ((length + lengthOfScanRsp) < BLE_MAX_ADV_SIZE && 
+             (_scan_rsp_data_idx < ARRAY_SIZE(_scan_rsp_data)))
+    {
+        fill_area = &_scan_rsp_data[_scan_rsp_data_idx];
+        _scan_rsp_data_idx++;
+    }
+    else
+    {
+        // Service data block is too large.
+        return BLE_STATUS_ERROR_PARAMETER;
+    }
+
+    if (fill_area)
+    {
+        fill_area->type = type;
+        fill_area->data = data;
+        fill_area->data_len = length;
+        
+        pr_info(LOG_MODULE_BLE, "ADV type %d Len - %d",type, length);
+    }
+    return BLE_STATUS_SUCCESS;
+}
+
+BLE_STATUS_T
 BLEDeviceManager::_advDataInit(void)
 {
-    uint8_t lengthTotal = 2; // Flags data length
+    BLE_STATUS_T ret = BLE_STATUS_SUCCESS;
+    // Clear the indexs
     _adv_data_idx = 0;
     _scan_rsp_data_idx = 0;
     
     /* Add flags */
     _adv_type = (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR);
-    _adv_data[_adv_data_idx].type = BT_DATA_FLAGS;
-    _adv_data[_adv_data_idx].data = &_adv_type;
-    _adv_data[_adv_data_idx].data_len = 1;
-    _adv_data_idx++;
+    ret = setAdvertiseData (BT_DATA_FLAGS, &_adv_type, sizeof(_adv_type));
     
-    if (_has_service_uuid) 
+    if (_has_service_solicit_uuid && 
+        (BLE_STATUS_SUCCESS == ret)) 
     {
         uint8_t type;
         uint8_t length;
         uint8_t *data = NULL;
         
-        pr_info(LOG_MODULE_BLE, "ADV Type-%d", _service_uuid.uuid.type);
-        if (BT_UUID_TYPE_16 == _service_uuid.uuid.type)
+        pr_info(LOG_MODULE_BLE, "ADV Type-%d", _service_solicit_uuid.uuid.type);
+        if (BT_UUID_TYPE_16 == _service_solicit_uuid.uuid.type)
         {
-            //UINT16_TO_LESTREAM(adv_tmp, uuid.uuid16);
-            data = (uint8_t *)&(((bt_uuid_16_t *)&_service_uuid)->val);
+            data = (uint8_t *)&(((bt_uuid_16_t *)&_service_solicit_uuid)->val);
```

Comment:
What's going here? Can't you just do this instead;

`data = (uint8_t *) &_service_solicit_uuid.val;`